src/adera/ShipResources.h

```diff
@@ -0,0 +1,175 @@
+ * "only" 4294967292 m^3.
+ *
+ */
+struct ShipResourceType
+{
+    // A short, unique, identifying name readable by both human and machine
+    std::string m_identifier;
+
+    // The full, screen-display name of the resource
+    std::string m_displayName;
+
+    // 1/(2^quanta) is the smallest representable quantity of this resource
+    uint8_t m_quanta;
+
+    // The mass (in kg) of 2^quanta units of this resource
+    float m_mass;
+
+    // The volume (in m^3) of 2^quanta units of this resource
+    float m_volume;
+
+    // The density of this resource (kg/m^3)
+    float m_density;
+};
+
+struct ShipResource
+{
+    osp::DependRes<ShipResourceType> m_type;
+    uint64_t m_quantity;
+};
+
+class MachineContainer;
+
+class SysMachineContainer
+    : public osp::active::SysMachine<SysMachineContainer, MachineContainer>
+{
+public:
+    SysMachineContainer(osp::active::ActiveScene& rScene);
+
+    static void update_containers(osp::active::ActiveScene& rScene);
+
+    static float resource_volume(ShipResourceType type, uint64_t quantity);
+    static float resource_mass(ShipResourceType type, uint64_t quantity);
+    static uint64_t resource_capacity(ShipResourceType type, float volume);
+
+    osp::active::Machine& instantiate(
+        osp::active::ActiveEnt ent,
+        osp::PrototypeMachine config,
+        osp::BlueprintMachine settings) override;
+
+    osp::active::Machine& get(osp::active::ActiveEnt ent) override;
+
+private:
+    
+    osp::active::UpdateOrderHandle_t m_updateContainers;
+}; // class SysMachineContainer
+
+class MachineContainer : public osp::active::Machine
+{
+    friend SysMachineContainer;
+
+public:
+    MachineContainer(float capacity, ShipResource resource);
+    MachineContainer(MachineContainer&& move) noexcept;
+    MachineContainer& operator=(MachineContainer&& move) noexcept;
+    ~MachineContainer() = default;
+
+    void propagate_output(osp::active::WireOutput *output) override;
+
+    osp::active::WireInput* request_input(osp::WireInPort port) override;
+    osp::active::WireOutput* request_output(osp::WireOutPort port) override;
+
+    std::vector<osp::active::WireInput*> existing_inputs() override;
+    std::vector<osp::active::WireOutput*> existing_outputs() override;
+
+private:
+    std::vector<osp::active::WireInput*> m_inputs;
+    std::vector<osp::active::WireOutput*> m_outputs;
+
+    float m_capacity;
+    ShipResource m_contents;
+}; // class MachineContainer
+
+/* MachineContainer */
+inline MachineContainer::MachineContainer(float capacity, ShipResource resource)
+    : Machine(true)
+    , m_capacity(capacity)
+    , m_contents(resource)
+{}
+
+inline MachineContainer::MachineContainer(MachineContainer&& move) noexcept
+    : Machine(std::move(move))
+    , m_capacity(std::move(move.m_capacity))
+    , m_contents(std::move(move.m_contents))
+{}
+
+inline MachineContainer& MachineContainer::operator=(MachineContainer&& move) noexcept
+{
+    m_enable = move.m_enable;
```

Comment:
Conceptually, is it still enabled after being moved from?

I recommend std:;exchange for situations where the moved-from object shouldn't be in an inconstant state.